Add authentication, dashboard, navigation, and filtering features

omarr78 opened the pull request.

omarr78 committed:
create login component layout

omarr78 committed:
add login component layout (HTML & CSS only)

omarr78 committed:
implement form validation with ReactiveForms

omarr78 committed:
feat(auth): implement authentication flow with API integration

- add ApiService for handling endpoints
- create UserCredentials interface
- use HttpClient to send requests
- add AuthService to authenticate users and store token in sessionStorage
- handle login backend errors (401, connection issues)
- create Dashboard component and redirect after login

omarr78 committed:
Add loading state and disable sign in button until API response

omarr78 committed:
minor html input validation update

Seifehab11-8 committed:
made the username accept any character other than spaces

Seifehab11-8 committed:
changed the error message for the spaces

omarr78 committed:
feat(dashboard): implement 'list all rooms' feature

- add dashboard component layout (HTML & CSS)
- add endpoint in ApiService
- create Room interface
- create and register HTTP interceptor in app.config.ts
- implement API call logic in dashboard component with dummy backend
- remove duplicate authentication files
- fix issues in authentication service

omarr78 committed:
integrate real backend for room listing

- replace dummy backend with real API + database
- update MeetingRoom interface to match backend response
- adjust dashboard HTML & TS to display real values

omarr78 committed:
Install SweetAlert2 package & integrate toast notification on successful login

omarr78 committed:
feat(guards): add AuthGuard for dashboard and NoAuthGuard for login

- prevent unauthenticated users from accessing dashboard
- prevent authenticated users from accessing login page

omarr78 committed:
add PageNotFound component with AuthGuard

omarr78 committed:
Add Navbar & logout

- create Navbar component
- include Navbar in app component and display only when user is logged in
- implement logout functionality inside Navbar
- show alert on logout
- add Navbar HTML & CSS layout
- adjust colors in other components for consistency

omarr78 committed:
fix home navigation & fix navbar css issue

omarr78 committed:
clean up commented code

omarr78 committed:
Add filter

- create filter component with HTML & CSS styling
- add filter button in dashboard to toggle filter component
- create reusable modal in UI
- define Equipment interface and add dummy equipment list
- include meeting room images in project assets
- update code to match backend API changes

mmahmoud07 commented on src/app/core/services/api/api.service.ts:
```diff
@@ -0,0 +1,22 @@
+})
+export class ApiService {
+
+  private apiUrl = 'http://localhost:8080';
```
It's better to read it from environment file instead of hard coding it

mmahmoud07 commented on src/app/dashboard/dashboard.component.ts:
```diff
@@ -0,0 +1,45 @@
+    this.api.getRooms().subscribe(
+      {
+        next: (response) => {
+          console.log(response.body);
```
no need to this `console.log`

mmahmoud07 commented on src/app/filter/filter.component.ts:
```diff
@@ -0,0 +1,54 @@
+
+  constructor(private modal:ModalComponent){}
+
+  equipments: Equip[] = [
+    { id: 1, name: "Projector or Large Display Screen / TV", isExist: false },
+    { id: 2, name: "Whiteboard", isExist: false },
+    { id: 3, name: "Flip Chart", isExist: false },
+    { id: 4, name: "Conference Phone (Speakerphone)", isExist: false },
+    { id: 5, name: "Video Conferencing System (Camera + Mic)", isExist: false },
+    { id: 6, name: "Wireless Presentation System", isExist: false },
+    { id: 7, name: "HDMI / VGA Cables & Adapters", isExist: false },
+    { id: 8, name: "Laser Pointer / Presenter Remote", isExist: false },
+    { id: 9, name: "Room Scheduling Display", isExist: false },
+    { id: 10, name: "Power Outlets & Charging Stations", isExist: false },
+    { id: 11, name: "Wi-Fi Access Point", isExist: false },
+    { id: 12, name: "Microphones", isExist: false },
+    { id: 13, name: "Speakers / Sound System", isExist: false },
+    { id: 14, name: "Lighting Control System", isExist: false },
+    { id: 15, name: "Air Conditioning / Climate Control", isExist: false },
+  ];
+
+
+  capacities:number[] = [10,15,25,35];
```
Why are these values hardcoded?

mmahmoud07 commented on src/app/filter/filter.component.ts:
Is there any action done by this component?

mmahmoud07 commented on src/app/guards/auth.guard.ts:
```diff
@@ -0,0 +1,20 @@
+    const isLoggedIn = sessionStorage.getItem(TOKEN) !== null;
+
+    if (isLoggedIn) return true;
+    this.router.navigate(['login']);
+    return false;
```
We can return a `UrlTree` instead of navigating and returning false.

mmahmoud07 commented on src/app/guards/no-auth.guard.ts:
```diff
@@ -0,0 +1,23 @@
+
+    if(isLoggedIn){
+        //  already logged in → redirect to dashboard
+        this.router.navigate(['dashboard']); 
+        return false;
```
Same, we can return a UrlTree instead of navigating and returning false.

mmahmoud07 commented on src/app/interceptor/token.interceptor.ts:
```diff
@@ -0,0 +1,15 @@
+  
+  const token = sessionStorage.getItem(TOKEN);
+  if (token) {
+    console.log('from interceptor', token);
```
no need for this `console.log()`

mmahmoud07 commented on src/app/core/services/authentication/authentication.service.ts:
It's better to manage token in a single place (generate, store, get, remove), and to use this service in other locations

mmahmoud07 commented on src/app/login/login.component.ts:
```diff
@@ -0,0 +1,88 @@
+              icon: "success",
+              title: "You logged in successfully."
+            });
+            console.log(response);
```
no need for this `console.log`

mmahmoud07 commented on src/app/login/login.component.ts:
```diff
@@ -0,0 +1,88 @@
+              this.errorMessage = 'Something went wrong. Please try again.';
+            }
+            this.isLoading = false;
+            console.log(this.errorMessage);
```
same, no need for this `console.log`

mmahmoud07 commented on src/app/nav-bar/nav-bar.component.ts:
```diff
@@ -0,0 +1,31 @@
+  }
+
+  logout(){
+    sessionStorage.removeItem(TOKEN);
```
move this logic to auth service

mmahmoud07 commented on src/app/login/login.component.ts:
```diff
@@ -0,0 +1,88 @@
+      this.auth.authenticate(user).subscribe(
+        {
+          next: (response) => {
+            sessionStorage.setItem(TOKEN, response.body.token);
```
move this logic to auth service

mmahmoud07 reviewed (commented):
please check the left comments

omarr78 committed:
add filter input form validation

omarr78 committed:
Dynamic equipment, Room filtering, and Validations

- Replaced hardcoded equipment with data from backend
- Added API request to fetch all equipment
- Created Equipment and EquipmentResponse models with isChecked property
- Implemented dashboard filtering using shared FilterService
- Made room capacity dynamic with input validation
- Moved all validation logic to shared validators.ts
- Added loading animation while fetching filtered rooms
- Added 'No Result Found' message when no rooms match filters
- Replaced getRooms API request with filteredRoom request using filtered data
- Cleared filter service on logout

omarr78 committed:
fix the loading logic in dashboard

mmahmoud07 commented on src/app/core/models/equip-response.ts:
```diff
@@ -0,0 +1,4 @@
+export interface EquipResponse {
+    equipment_id:number,
```
Let's use camelCase convention

mmahmoud07 commented on src/app/core/models/equip.ts:
```diff
@@ -1,5 +1,14 @@
-    id: number,
-    name: string,
-    isExist: boolean
+    equipment_id: number,
```
Let's use camelCase convention and void mixing different naming conventions

mmahmoud07 commented on src/app/core/models/filter-request.ts:
```diff
@@ -0,0 +1,37 @@
+    let startTime: Date | null = null;
+    let endTime: Date | null = null;
+
+    if (filter?.date) {
+
+        if (filter?.startTime !== null) {
+            startTime = new Date(filter?.date);
+            startTime?.setHours(filter?.startTime, 0, 0, 0);
+        }
+
+        if (filter?.endTime !== null) {
+            endTime = new Date(filter?.date);
+            endTime?.setHours(filter?.endTime, 0, 0, 0);
+        }
+    }
```
What are we trying to set here? could you please show an example?

mmahmoud07 commented on src/app/core/models/filter.ts:
```diff
@@ -0,0 +1,9 @@
+    startTime: number | null,
+    endTime: number | null,
+    capacity: number | null,
+    equipmentTypes: Equip[]
```
Since this is holding the types only, we can make it of `String` type instead of `Equip`

mmahmoud07 commented on src/app/core/services/api/api.service.ts:
```diff
@@ -16,7 +17,36 @@ export class ApiService {
+      if (params.capacity !== null) {
+        httpParams = httpParams.set('capacity', params.capacity.toString());
+      }
+      if (params.equipmentTypes && params.equipmentTypes.length > 0) {
```
Can't we do `params.equipmentTypes?.length`?

mmahmoud07 commented on src/app/dashboard/dashboard.component.ts:
```diff
@@ -1,45 +1,94 @@
+    return max;
+  }
+
+  formatTime(hour: number | null | undefined): string {
+    if (hour == null || hour == undefined) return '';
+    const suffix = hour >= 12 ? 'PM' : 'AM';
+    const displayHour = hour > 12 ? hour - 12 : hour;
+    return `${displayHour}:00 ${suffix}`;
+  }
```
I think there is another way to define data formatters to be used in the html

mmahmoud07 reviewed (commented):
I've left more comments
